ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java

```diff
@@ -435,30 +444,39 @@ private void testFailureDetectorNotification(OperationHelper opHelper, NetworkCl
     ReplicaId replicaIdToFail =
         indexToFail == -1 ? null : ((RouterRequestInfo) allRequests.get(indexToFail)).getReplicaId();
     for (RequestInfo requestInfo : allRequests) {
+      ResponseInfo responseInfo;
       if (replicaIdToFail != null && replicaIdToFail.equals(((RouterRequestInfo) requestInfo).getReplicaId())) {
-        mockSelectorState.set(MockSelectorState.DisconnectOnSend);
+        responseInfo = new ResponseInfo(requestInfo,
+            createConnectionUnavailableError ? NetworkClientErrorCode.ConnectionUnavailable
+                : NetworkClientErrorCode.NetworkError, null);
       } else {
         mockSelectorState.set(MockSelectorState.Good);
+        List<RequestInfo> requestInfoListToSend = new ArrayList<>();
+        requestInfoListToSend.add(requestInfo);
+        List<ResponseInfo> responseInfoList;
+        loopStartTimeMs = SystemTime.getInstance().milliseconds();
+        do {
+          if (loopStartTimeMs + AWAIT_TIMEOUT_MS < SystemTime.getInstance().milliseconds()) {
+            Assert.fail("Waited too long for the response.");
+          }
+          responseInfoList = networkClient.sendAndPoll(requestInfoListToSend, 10);
+          requestInfoListToSend.clear();
+        } while (responseInfoList.size() == 0);
+        responseInfo = responseInfoList.get(0);
       }
-      List<RequestInfo> requestInfoListToSend = new ArrayList<>();
-      requestInfoListToSend.add(requestInfo);
-      List<ResponseInfo> responseInfoList;
-      loopStartTimeMs = SystemTime.getInstance().milliseconds();
-      do {
-        if (loopStartTimeMs + AWAIT_TIMEOUT_MS < SystemTime.getInstance().milliseconds()) {
-          Assert.fail("Waited too long for the response.");
-        }
-        responseInfoList = networkClient.sendAndPoll(requestInfoListToSend, 10);
-        requestInfoListToSend.clear();
-      } while (responseInfoList.size() == 0);
-      opHelper.handleResponse(responseInfoList.get(0));
+      opHelper.handleResponse(responseInfo);
     }
     futureResult.get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
     if (indexToFail == -1) {
       Assert.assertEquals("Successful notification should have arrived for replicas that were up",
           opHelper.requestParallelism, successfulResponseCount.get());
       Assert.assertEquals("Failure detector should not have been notified", 0, failedReplicaIds.size());
       Assert.assertFalse("There should be no notifications of any other kind", invalidResponse.get());
+    } else if (createConnectionUnavailableError) {
```

Comment:
what is the difference b/w this code block and the `if(indexToFail == -1)` block?